File: webdav/infra/path.ts
import { IncomingMessage } from 'http'
import path from 'path'
import { DATA_DIRECTORY } from '@/config/config'

export function getFilePath(url: string) {
  return path.join(DATA_DIRECTORY, decodeURIComponent(url))
}

export function getDestinationPath(req: IncomingMessage) {
  const destinationHeader = req.headers.destination as string
  if (!destinationHeader) {
    return null
  }
  // Check if the destination header is a full URL
  if (
    destinationHeader.startsWith('http://') ||
    destinationHeader.startsWith('https://')
  ) {
    const url = new URL(destinationHeader)
    return path.join(DATA_DIRECTORY, decodeURIComponent(url.pathname))
  } else {
    /* Extract the path from the destination header */
    const startIndex =
      destinationHeader.indexOf(req.headers.host) + req.headers.host.length
    const value = destinationHeader.substring(startIndex)
    return path.join(DATA_DIRECTORY, decodeURIComponent(value))
  }
}

File: webdav-go/helper/path.go
package helper

import (
	"net/http"
	"net/url"
	"path"
	"strings"
)

func GetTargetPath(req *http.Request) string {
	destination := req.Header.Get("Destination")
	if destination == "" {
		return ""
	}
	/* Check if the destination header is a full URL */
	if strings.HasPrefix(destination, "http://") || strings.HasPrefix(destination, "https://") {
		parsedURL, err := url.Parse(destination)
		if err != nil {
			return ""
		}
		return parsedURL.Path
	}
	/* Extract the path from the destination header */
	startIndex := strings.Index(destination, req.Host) + len(req.Host)
	if startIndex < len(req.Host) {
		return ""
	}
	return destination[startIndex:]
}

func Dirname(value string) string {
	trimmedValue := strings.TrimSuffix(value, "/")
	return path.Dir(trimmedValue)
}
